bwt/bwt.go

```diff
@@ -0,0 +1,227 @@
+		}
+	}
+	panic("Unable to find the corresponding orginal positiong for a character in the original sequence in the suffix array. This should not be possible and indicates a malformed BWT.")
+}
+
+// lfSearch LF Search- Last First Search.
+// Finds the valid range within the BWT index where the provided pattern is possible.
+// If the final range is <= 0, then the pattern does not exist in the original sequence.
+func (bwt BWT) lfSearch(pattern string) interval {
+	searchRange := interval{start: 0, end: bwt.getLenOfOriginalString()}
+	for i := 0; i < len(pattern); i++ {
+		if searchRange.end-searchRange.start <= 0 {
+			return interval{}
+		}
+
+		c := pattern[len(pattern)-1-i]
+		skip, ok := bwt.lookupSkipByChar(c)
+		if !ok {
+			return interval{}
+		}
+		searchRange.start = skip.openEndedInterval.start + bwt.lastCoulmn.Rank(c, searchRange.start)
+		searchRange.end = skip.openEndedInterval.start + bwt.lastCoulmn.Rank(c, searchRange.end)
+	}
+	return searchRange
+}
+
+// lookupSkipByChar looks up a skipEntry by its character in the First Coulmn
+func (bwt BWT) lookupSkipByChar(c byte) (entry skipEntry, ok bool) {
+	for i := range bwt.firstColumnSkipList {
+		if bwt.firstColumnSkipList[i].char == c {
+			return bwt.firstColumnSkipList[i], true
+		}
+	}
+	return skipEntry{}, false
+}
+
+// lookupSkipByOffset looks up a skipEntry based off of an
+// offset of the Fist Coulmn of the BWT.
+func (bwt BWT) lookupSkipByOffset(offset int) skipEntry {
+	if offset > bwt.getLenOfOriginalString()-1 {
+		msg := fmt.Sprintf("offset [%d] exceeds the max bound of the BWT [%d]", offset, bwt.getLenOfOriginalString()-1)
+		panic(msg)
+	}
+	if offset < 0 {
+		msg := fmt.Sprintf("offset [%d] exceeds the min bound of the BWT [0]", offset)
+		panic(msg)
+	}
+
+	for i := range bwt.firstColumnSkipList {
+		if bwt.firstColumnSkipList[i].openEndedInterval.start <= offset && offset < bwt.firstColumnSkipList[i].openEndedInterval.end {
+			return bwt.firstColumnSkipList[i]
+		}
+	}
+	panic("figure out what to do here")
+}
+
+func (bwt BWT) getLenOfOriginalString() int {
+	return bwt.firstColumnSkipList[len(bwt.firstColumnSkipList)-1].openEndedInterval.end
+}
+
+type interval struct {
+	start int
+	end   int
+}
+
+type skipEntry struct {
+	char byte
+	// openEndedInterval start is inclusive and end is exclusive
+	openEndedInterval interval
+}
+
+// New returns a BWT of the provided sequence
+// The provided sequence must not contain the nullChar
+// defined in this package. If it does, New will return
+// an error.
+func New(sequence string) (BWT, error) {
+	if strings.Contains(sequence, nullChar) {
+		return BWT{}, fmt.Errorf("Provided sequence contains the nullChar %s. BWT cannot be constructed", nullChar)
+	}
+
+	sequence += nullChar
+
+	prefixArray := make([]string, len(sequence))
+	for i := 0; i < len(sequence); i++ {
+		prefixArray[i] = sequence[len(sequence)-i-1:]
+	}
+
+	// TODO: at the time of writing, the nullChar is 0, this is to ensure correctness in most cases.
```

Comment:
I think this may be a good idea? Is this why I've been seeing "$" in other implementations?

Reply:
I think I'll just give the custom sorting a shot on the string to ensure that the nullChar is always the least no matter what so we can put back the $.

Reply:
Done, LMK what you think. nullChar could be configurable as a functional option now if users need to have $ as a part of their alphabet for the index they are throwing in to the BWT for whatever reason.